src/utilities.jl

```diff
@@ -25,36 +25,83 @@ function overlaps(ts::Vararg{Vector, N}) where {N}
     ret
 end
 
+"""
+    sorted_unique_merge(a, b) -> (c, idx_a, idx_b)
 
-function sorted_unique_merge(a::Vector, b::Vector)
-
-		i, na, j, nb = 1, length(a), 1, length(b)
-		c = similar(a, 0)
+Merge two arrays of sorted unique elements a and b to new array c,
+and calculate the indexes idx_a and idx_b mapping each element in a and b to their new locations in c.
+"""
+function sorted_unique_merge(::Type{IndexType}, a::Vector, b::Vector) where {IndexType}
+    i, na, j, nb = 1, length(a), 1, length(b)
+    c = similar(a, length(a) + length(b))
+    idx_a = Vector{IndexType}(length(a))
+    idx_b = Vector{IndexType}(length(b))
+    k = 1
+    @inbounds while (i <= na) && (j <= nb)
+        if a[i] < b[j]
+            c[k] = a[i]
+            idx_a[i] = k
+            i += 1
+        elseif a[i] > b[j]
+            c[k] = b[j]
+            idx_b[j] = k
+            j += 1
+        else
+            c[k] = a[i]
+            idx_a[i] = k
+            idx_b[j] = k
+            i += 1
+            j += 1
+        end
+        k += 1
+    end
+    @inbounds while i <= na
+        c[k] = a[i]
+        idx_a[i] = k
+        i += 1
+        k += 1
+    end
+    @inbounds while j <= nb
+        c[k] = b[j]
+        idx_b[j] = k
+        j += 1
+        k += 1
+    end
+    resize!(c, k - 1)
+    return c, idx_a, idx_b
+end
 
-		while (i <= na) && (j <= nb)
-				if a[i] < b[j]
-						push!(c, a[i])
-						i += 1
-				elseif a[i] > b[j]
-						push!(c, b[j])
-						j += 1
-				else
-						push!(c, a[i])
-						i += 1
-						j += 1
-				end #if
-		end
+"""
+    insertbyidx!(dst, src, dstidx, coloffset = 0)
 
-		append!(c, a[i:end])
-		append!(c, b[j:end])
+For each column in src, insert elements from src[i, column] to dst[dstidx[i], column + coloffset].
+"""
+function insertbyidx!(dst::AbstractArray, src::AbstractArray, dstidx::Vector, coloffset::Int = 0)
+    for c in 1:size(src, 2)
+        cc = c + coloffset
+        for i in 1:length(dstidx)
+            @inbounds dst[dstidx[i], cc] = src[i, c]
+        end
+    end
+    nothing
+end
 
-    return c
+"""
+    insertbyidx!(dst, src, dstidx, srcidx)
 
-end  # sorted_unique_merge
+For each column in src, insert elements from src[srcidx[i], column] to dst[dstidx[i], column].
+"""
+function insertbyidx!(dst::AbstractArray, src::AbstractArray, dstidx::Vector, srcidx::Vector)
```

Comment:
I think this function can replace by 
```julia
broadcast_setindex!(dst, broadcast_getindex(src, srcidx), dstidx)
```

which is faster in your use case?

Reply:
```Julia
using BenchmarkTools
using TimeSeries
dst = zeros(1_000_000)
src = zeros(1_000_000)
dstidx = [1:length(dst)...]
srcidx = [1:length(dst)...]
@btime broadcast_setindex!(dst, broadcast_getindex(src, srcidx), dstidx)
@btime TimeSeries.insertbyidx!(dst, src, dstidx, srcidx)
```
  9.486 ms (8 allocations: 7.63 MiB)
  2.208 ms (0 allocations: 0 bytes)

Reply:
I would love for the core language and library features to work optimally, so worth revisiting this in the future.